Improve: provide alternative representations of most control characters

SlySven opened the pull request:
For some MUDs that want to use the CP437 encoding they may also want the C0 control characters (those in the range 0-31 and at 127) to yield visible glyphs.  Whilst there are some control characters that must retain their effects (say Tab, Carriage Return, Line Feed, Escape) it could be useful to show a visible representation for the remainder instead of letting the code that paints text onto the `TConsole`s of the profile fail and moan about being asked to display non-printable Unicode code-points.

This PR provides two alternatives:
* the glyphs that the CP437 encoding seems to provide for those control characters see: https://en.wikipedia.org/wiki/Code_page_437#Character_set
* the Unicode "Control Pictures" found at U+2400 to U+2421

The setting is saved on a per profile basis.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Other info (issues closed, discussion etc)

#### Release post highlight
Improvement, can now visualise most "control characters" which is useful if a MUD is using them as normal characters (those trying to reproduce the PC "OEM Font" that CP437 is often used for) - or for debugging purposes!

SlySven committed:
Improve: provide alternative representations of most control characters

For some MUDs that want to use the CP437 encoding they may also want the
C0 control characters (those in the range 0-31 and at 127) to yield visible
glyphs.  Whilst there are some control characters that must retain their
effects (say Tab, Carriage Return, Line Feed, Escape) it could be useful to
show a visible representation for the remainder instead of letting the
code that paints text onto the `TConsole`s of the profile fail and moan
about being asked to display non-printable Unicode code-points.

This PR provides two alternatives:
* the glyphs that the CP437 encoding seems to provide for those control
characters see: https://en.wikipedia.org/wiki/Code_page_437#Character_set
* the Unicode "Control Pictures" found at U+2400 to U+2421

The setting is saved on a per profile basis.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: undo a last minute change to "clean up" something that doesn't work


This is because there is no `QVection<QString>::append(QChar)` method, so
a `(enum) QChar::SpecialCharacter` has to be explicitly converted to a
`QString` *before* that can be appended to the container. I had forgotten
that and removed the intermediate `QString(...)` in what I thought was a
last minute improvement!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: keep CodeSpell bot happy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
CodeFactor is not happy I have extended `(int) TTextEdit::drawGraphemeBackground(...)` to include a nested `switch` - the outer one contains two (one for each new "mode") plus the existing code for the existing behaviour) - I suppose it might be prettier to refactor each one to a separate method - which should be `inline`d as it would only be used one in the application. The other thing is added a few lines to the Host export method in the `XMLExport` class which is big - but there is no way around that as it is currently structured.

SlySven commented on src/Host.cpp:
```diff
@@ -3757,3 +3757,15 @@ void Host::setupIreDriverBugfix()
+
+void Host::setControlCharacterMode(const int mode)
+{
+    if (mode < 0 && mode > 2) {
```
:man_facepalming:

SlySven committed:
Refactor: make a method smaller by extracting two chunks to helpers

This will hopefully mollify CodeFactor a little!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: a combinational logic error on my part

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Nice 👍 

How does it relate to https://github.com/Mudlet/Mudlet/issues/4332 ?

vadi2 commented on src/Host.h:
```diff
@@ -773,6 +778,18 @@ private slots:
+    //   EN: https://en.wikipedia.org/wiki/Code_page_437
+    //   DE: https://de.wikipedia.org/wiki/Codepage_437
+    //   RU: https://ru.wikipedia.org/wiki/CP437
+    int mControlCharacterHandlingMode;
```
Would `mControlCharacterMode` work just as well?

Also, could we make it an `enum` not to use a magic number? We use magic numbers for modules/packaging and it is confusing.

SlySven replied:
Yeah, that makes sense, only issue with `enum`s is that IIRC sometimes it can take more effort to read/write them from/to a `QDataStream` as Qt/compiler gets unhappy about converting between a `enum` and the underlying integer form. I'll give it a go and see if I can find the magic invocations to use...

SlySven committed:
Revise: correct an off-by-one copy-paste error

I got part of the range of the OEM alternative characters off-set by one
when I copied and replaced the list of characters from the "Control
Picture" one.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> How does it relate to #4332 ?

If someone is getting `<CR>`s when they shouldn't be and they have one of the alternatives turned on, then, ideally, they ought to see a `␍` or a `♪` instead - though for that particular character it depends on how it gets introduced into the application as, by some routes, it ought to be swallowed up with a following <NUL> or <LF> as an End-of-Line...

vadi2 commented:
And by default, it gets ignored as it was before `␍` was added?

SlySven commented:
> And by default, it gets ignored as it was before ␍ was added?

Not at the moment, but specifically about that character I am also looking at how we treat line endings and I want to investigate whether we should record the end-of-lines in the `(QStringList) TBuffer::lineBuffer` (but hide them from the trigger engine and other code that *doesn't* need to know - but make them available for things that should {like the text analyser}). That should take care of them...

SlySven committed:
Revise: keep Clang-Tidy bot happy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Been asking for this for a year... it's a control character, it was changed to everyone's dissatisfaction a year ago, need to see it fixed now.

SlySven commented:
Okay - am looking at that right now to see if (as a supplement to this PR as handling the replacement of undraw-able characters is right in the same area) I can keep the `TTextEdit` text painting code working whilst NOT trying to draw graphemes that are "undraw-able"...

vadi2 commented:
thank you!

SlySven committed:
Enhance: don't use Replacement Character for unprintable characters

This is a bit experimental but seems to work.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
We should apply postel's law to data coming from the game: be accepting as we can. This also means we should not assume every player has the technical context to understand what are encoding issues, and thus we should not show them by default -

![image](https://user-images.githubusercontent.com/110988/140635875-1806f62a-3829-400e-9d20-70193ddd4e47.png)

SlySven commented:
Regarding the last message - you have a point - however this is also about providing printable replacements for (most) control characters - and this *is needed* I think for at least one MUD (Medievia `medievia.com:4000` ) that is sending a control character 0x18 (24 dec) IIRC and is expecting a particular character (a vertical arrow) at that position as it is (or trying to) reproduce/ing the OEM Font characters that MSDos PCs used many years ago.

SlySven committed:
Merge branch 'development' into Improve_enhance_provideAlterativeRepresentationsOfMostControlCharacters

Conflicts resolved in:
* src/Host.cpp
* src/TTextEdit.h
* src/dlgProfilePreferences.cpp
* src/dlgProfilePreferences.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/TTextEdit.cpp:
```diff
@@ -468,29 +468,138 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen, in
     TChar::AttributeFlags attributes = charStyle.allDisplayAttributes();
-    auto textRect = QRect(mFontWidth * cursor.x(), mFontHeight * cursor.y(), mFontWidth * charWidth, mFontHeight);
+    QRect textRect;
+    if (charWidth > 0) {
```
```suggestion
    if (Q_LIKELY(charWidth > 0)) {
```

vadi2 commented:
Can confirm default of no replacement character works well!

The `General` page is a bit more horrifying to look at now though - 
![image](https://user-images.githubusercontent.com/110988/149285484-1360985d-30cc-4ebd-8fae-84800bcaf338.png)

Idea on how to fix @SlySven @Kebap ?

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -2959,7 +2959,12 @@ Some data loss is likely - please mention this problem to the game admins.)", co
             } else {
                 if (ch == TN_BELL) {
-                    // flash taskbar for 3 seconds on the telnet bell
+                    // Flash taskbar for 3 seconds on the telnet bell, note
+                    // by processing it here rather than in the TTextEdit class
+                    // it is not possible to fake/test it with a Lua
+                    // feedTriggers(...) call - OTOH doing it there would make
+                    // a beep every time the screen was refreshed!
+                    // TODO: https://github.com/Mudlet/Mudlet/issues/5836 - provide option to actually make a (void) QApplication::beep() or a user-selected sound (different for each profile) and/or instead of the visual alert
```
I'm not sure what value this adds, it's a lot of text but for what purpose? Does make code harder to read though.

SlySven replied:
I wanted to document that \<BEL>included in a `feedTriggers(...)`call will NOT trigger this code...

Kebap commented:
> how to fix

- use drop-down menu instead
- move from General to Main Display
- rephrase:

Display control characters?
- No
- With Unicode
- Like CP437

Note how Unicode and CP437 are actual names and capitalized unlike Control Characters..


> this is needed I think

Did you confirm with actual players?

vadi2 commented:
That's a great idea 👍

SlySven committed:
Revise: reposition control character replacement options

Move it to the third tab and put the control in the Display options
`QGroupBox`.

Also fixup the *tab-order* for these new items and thy enabling and
disabling of the controls when there is or isn't a profile loaded.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Note how Unicode and CP437 are actual names and capitalized unlike Control Characters.

If you do a search on the Internet for "unicode" it does tend to be capitalised, similarly searching for "unicode control pictures" the 'C' tends to be though the 'P' less so. Actually, even as I type this text, using a lower case 'u' as the initial letter for Unicode is being flagged as a spelling error!

As for the point about it taking up too much space on the "General" tab I have moved it to the bottom groupbox on the third tab and arranged for the tab stops to be in the right order and for it to be correctly disabled/enabled if there isn't/is a profile loaded (as it is a per profile option).

vadi2 commented:
I believe Kebap meant that it should be just 'control characters'. I agree, it does not warrant capitalisation.

How about we switch the radio box to a combobox to save space? This option will be hardly ever used, it does not require the quick accessibility of radio buttons - but on the other hand we do have a space issue, so using least possible is better.

SlySven commented:
https://github.com/Mudlet/Mudlet/pull/5573#issuecomment-1012798279 

TBH Radio buttons are the logical UI choice for one of a number of mutually exclusive options as that shows what the options all are without having to examine a combo-box to see them. I'd prefer them but I suppose I can rework the control to be one of the latter... 🙁

SlySven committed:
Revise: change from radio buttons to combobox and use enum

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Improve player-facing text

vadi2 commented:
Thank for doing the actual heavy lifting on this, not only does it solve the unprintable characters thing but also makes custom fonts work great!

We need to document very well how can someone create a Mudlet package with a custom font and make the fancy characters display right, so people aren't left in the dark about this amazing feature -

***Edit: removed screenshot of a MUD's custom font - the details were shown to SlySven privately and it was not clear that the information was to be widely disseminated.*** 

I improved the texts in the UI a bit to make them friendlier to translators and players. Before -

![Screenshot 2022-01-16 8 43 40 AM](https://user-images.githubusercontent.com/110988/149651920-1a18009d-b143-4769-8906-e885a000db92.png)

Now -

![Screenshot 2022-01-16 8 52 27 AM](https://user-images.githubusercontent.com/110988/149651928-48043146-4e37-4f36-90b7-5c2f958f8e49.png)

SlySven commented:
I've disabled the auto-merge as, because there are multiple commits, it will just concatenate them all into one and leave a single messy commit message. I'll watch the CI/CB and manually merge it with a cleaned up message in the next half hour or so once the Windows build has completed...
